### DISCUSSION TITLE
One SyncAgent, multiple clients, is this possible?

### DISCUSSION THREAD

**Discussion by badsyntax** [Mimetis/Dotmim.Sync#833] [Q&A] [answered]
Hi. I'm evaluating using this library for data sync. 

We have a multi-tenant application, where each tenant/client has their own database. We have one API service that serves all tenants.

We have some common data that we'd like to sync into each tenant database.

I've had a read through the [Multiple Client overview](https://dotmimsync.readthedocs.io/HowDoesItWorks.html#multiple-clients-overview) but it seems those docs are targeted towards clients having their own servers. Conceptually we only have 1 server. The docs mention `And each client will have a sync agent, responsible of the sync process`. We can have `N` tenant databases, so we need to be able to dynamically configure the sync.

Ideally we'd only like one sync agent, that can sync to multiple clientProviders. Is this possible? If not, do you have any advice on how to sync to a non-static list of multiple clients? Thanks in advance.

**Comment by Mimetis** [comment, marked as answer]
> Conceptually we only have 1 server

That's good. 1 server and multiple clients. Classic scenario I guess

> We can have N tenant databases

On the client side ?

> Ideally we'd only like one sync agent, that can sync to multiple clientProviders

A sync agent ist just an object who knows the server and the client.
You can create as many SyncAgent as you want, on your local client.

**Reply by badsyntax** [reply]
>> We can have N tenant databases

> On the client side ?

Yes. There's only one "server" db that holds "common data". Each client/tenant has their own database. We want to sync the common data to each client/tenant database. 

>> Ideally we'd only like one sync agent, that can sync to multiple clientProviders

> A sync agent ist just an object who knows the server and the client.
> You can create as many SyncAgent as you want, on your local client.

Great, thank you. I'll create a POC and let you know if I have any difficulties.

**Reply by badsyntax** [reply]
Just want to say everything is working out great, and thanks for the useful library, it's solved some problems we had.

**Comment by VagueGit** [comment]
We use the [token in the HttpClient header](https://dotmimsync.readthedocs.io/WebSecurity.html?highlight=HTTPClient#client-side) to identify the user. Then configure the connection string to the server database in `SyncController.Post`

**Reply by badsyntax** [reply]
Thanks, I'll explore using the web proxy to see if i'll make this easier.

**Reply by AndyVu2311** [reply]
I got the same requirement and i'm working on it. Can you show more detail of the solution @VagueGit.

**Comment by VagueGit** [comment]
On the client we use this overload where the headers of httpClient carry some token that identifies the user
`WebRemoteOrchestrator webRemoteOrchestrator = new WebRemoteOrchestrator(serviceUri, client: httpClient, maxDownladingDegreeOfParallelism: 4);` Refer [Http Mode](https://dotmimsync.readthedocs.io/Orchestrators.html#http-mode)

The SyncController reads the headers to identify the user and configures the connection string for that user like this
`webServerAgent.Provider.ConnectionString = ...;` Refer [Server Side](https://dotmimsync.readthedocs.io/Web.html#server-side)
